packages/@aws-cdk/aws-ec2/lib/vpc-flow-logs.ts

```diff
@@ -0,0 +1,240 @@
+import iam = require('@aws-cdk/aws-iam');
+import logs = require('@aws-cdk/aws-logs');
+import s3 = require('@aws-cdk/aws-s3');
+import { Construct, IResource, Resource } from '@aws-cdk/core';
+import { CfnFlowLog } from './ec2.generated';
+import { ISubnet, IVpc } from './vpc';
+
+export interface IVpcFlowLog extends IResource {
+  /**
+   * The Id of the VPC Flow Log
+   */
+  readonly flowLogId: string;
+}
+
+/**
+ * The type of VPC traffic to log
+ */
+export enum VpcFlowLogTrafficType {
+  /**
+   * Only log accepts
+   */
+  ACCEPT = 'ACCEPT',
+
+  /**
+   * Log all requests
+   */
+  ALL = 'ALL',
+
+  /**
+   * Only log rejects
+   */
+  REJECT = 'REJECT'
+}
+
+/**
+ * The type of resource to create the flow log for
+ */
+enum VpcFlowLogResourceType {
+  NETWORK_INTERFACE = 'NetworkInterface',
+  SUBNET = 'Subnet',
+  VPC = 'VPC'
+}
+
+/**
+ * The destination type for the flow log
+ */
+export enum VpcFlowLogDestinationType {
+  /**
+   * CloudWatch logs destination
+   */
+  CLOUD_WATCH_LOGS = 'cloud-watch-logs',
+
+  /**
+   * S3 destination
+   */
+  S3 = 's3'
+}
+
+/**
+ * Options for the type of resource you
+ * are creating the flow log for
+ */
+export interface ResourceTypeOptions {
```

Comment:
The "union-like" pattern will fit this better. IIUC these are all mutually exclusive. Eventually you want the API to look like this:

```ts
new VpcFlowLog(this, 'flow', {
  resourceType: VpcFlowLogResourceType.fromSubnet(subnet)
});
```

To implement this, the pattern we've been using is:

```ts
export abstract class VpcFlowLogResourceType {
  public static fromSubnet(subnet: ISubnet): VpcFlowLogResourceType {
    return {
      resourceType: 'SUBNET',
      resourceId: subnet.subnetId
    };
  }

  public abstract resourceType: string;
  public abstract resourceId: string;
}
```

Reply:
I think I got this implemented, let me know what you think.